sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala

```diff
@@ -128,6 +128,65 @@ private[sql] object ArrowConverters extends Logging {
     }
   }
 
+  private[sql] def toArrowBatchIterator(
+      rowIter: Iterator[InternalRow],
+      schema: StructType,
+      maxRecordsPerBatch: Int,
+      timeZoneId: String,
+      context: TaskContext): Iterator[(Array[Byte], Long, Long)] = {
+    val arrowSchema = ArrowUtils.toArrowSchema(schema, timeZoneId)
+    val allocator = ArrowUtils.rootAllocator.newChildAllocator(
+      "toArrowBatchIterator", 0, Long.MaxValue)
+
+    val root = VectorSchemaRoot.create(arrowSchema, allocator)
+    val unloader = new VectorUnloader(root)
+    val arrowWriter = ArrowWriter.create(root)
+
+    if (context != null) { // for test at driver
+      context.addTaskCompletionListener[Unit] { _ =>
+        root.close()
+        allocator.close()
+      }
+    }
+
+    new Iterator[(Array[Byte], Long, Long)] {
+
+      override def hasNext: Boolean = rowIter.hasNext || {
+        root.close()
+        allocator.close()
+        false
+      }
+
+      override def next(): (Array[Byte], Long, Long) = {
+        val out = new ByteArrayOutputStream()
+        val writeChannel = new WriteChannel(Channels.newChannel(out))
+
+        var rowCount = 0L
+        var estimatedSize = 0L
+        Utils.tryWithSafeFinally {
+          while (rowIter.hasNext && (maxRecordsPerBatch <= 0 || rowCount < maxRecordsPerBatch)) {
+            val row = rowIter.next()
+            arrowWriter.write(row)
+            rowCount += 1
+            estimatedSize += SizeEstimator.estimate(row)
+          }
+          arrowWriter.finish()
+          val batch = unloader.getRecordBatch()
+
+          MessageSerializer.serialize(writeChannel, arrowSchema)
```

Comment:
Instead of turning each written result into a separate IPC stream, why not make the driver send the schema, and then stream back the record batches? I am not 100% how hard it would be to reassemble this on the python side, on the scala side it would be fairly straightforward.

Reply:
i had tried to split schema and record batches, but didn't find an easy way to do this. Then made this change as per @grundprinzip and @HyukjinKwon  suggestion

Reply:
The problem is that we carry Arrow record batch within each protobuf message. So, schema of each recordbatch has to be sent in order to read each protobuf message.

Reply:
Right now the client simply does a Pandas Union of all the batches coming from the server. The benefit is that we don't have to wait to serialize some data to the user. 

Logically it makes sense to send the schema only once, but right now my suggestion would be to keep it like this to make the default consumption easier. Every Arrow batch we send is a fully contained Arrow IPC stream.

Reply:
sgtm